### PR TITLE
Use `Renderable` trait to print to html output

### PR DESCRIPTION
**Pull request by Maaarcocr**
# why? 

it should fix https://github.com/Pitasi/rscx/issues/4

Currently whenever we print a value to the output buffer we assume that it's always going to implement `Display`. This works, but it has few drawbacks: 

- it's not implemented on `Option`
- If you want to implement it on some custom type/`Option` you end up implementing a fairly largely used trait, which not only specifies how stuff should be displayed in general, but also provides a `to_str()` implementation

Especially the second point is problematic (imo). I've worked in codebases where `Display` was relayed upon in a similar way and it ended up being a mess because people also started using `to_str` on types that should have not been transformed on strings and so on. 

Having our own trait I think is much cleaner and we can make it very similar to `Display`. 

# how? 

We add a new trait `Renderable` which is also implemented for a bunch of basic types. We also add a new type, `FormatWrapper` which implements `Display` on anything that implements `Renderable`. This is necessary as we still want to do `format!("...", x, y, z)` where `x,y,z` are the `out.values` that we have collected in the macro expansion. `format!` requires `Display` to work, so we use this (I think) zero cost abstraction to make the type system happy. 

# additional info

this is a breaking change, but this is pre 1.0.0, so I think it's fine

# lil easter egg

I've also made it such you do not need to import `EscapeAttribute` when using the macro expands to an expression that needs it, as we can fully qualify the method call in the macro expansion.

### CODE REVIEW

**Comment by Pitasi** [comment]
I see something around -30-50% penalty in the "rscx_small_fragment" benchmark. I did some initial investigations (see flamegraphs) and I came to conclusion that it must have something to do with traits and dynamic dispatch. Calls to the `render()` fn cannot be inlined.

[flamegraph before this pr](https://github.com/Pitasi/rscx/assets/15048443/9609f48f-9834-4749-9b95-290de7d31642)
[flamegraph after this pr](https://github.com/Pitasi/rscx/assets/15048443/0b0267df-69ca-4401-a8c4-6e8073348655)

I would have implemented this PR in pretty much the same way so thank you for your contribution. I'm keen to merge this anyway, but I want to experiment with it a bit more to see if we can gain back (at least some) performance.

File: rscx/src/renderable.rs [hunk 1]
```diff
@@ -0,0 +1,124 @@
+use std::{fmt, convert::Infallible, env::VarError, io::ErrorKind, net::{IpAddr, SocketAddr, Ipv4Addr, Ipv6Addr, SocketAddrV4, SocketAddrV6}, sync::{mpsc::{RecvTimeoutError, TryRecvError}, Arc}, num::{NonZeroI8, NonZeroI16, NonZeroI32, NonZeroI64, NonZeroI128, NonZeroIsize, NonZeroU8, NonZeroU16, NonZeroU32, NonZeroU64, NonZeroU128, NonZeroUsize}, borrow::Cow, ops::Deref, rc::Rc};
+
+pub trait Renderable {
```
Comment (thomaseizinger) [inline rscx/src/renderable.rs:3]: Maybe just call this `Render`? That is what most traits in stdlib do, e.g. `From`, `ToString`, `ToOwned`, ..

File: rscx/src/format_wrapper.rs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+pub struct FormatWrapper<T: Renderable> {
+    inner: T,
+}
+
+impl<T: Renderable> FormatWrapper<T> {
+    pub fn new(inner: T) -> Self {
+        Self { inner }
+    }
+}
```
Comment (thomaseizinger) [inline rscx/src/format_wrapper.rs:13]: Nit: Don't think it matters much for this PR but trait-bounds on structs or `impl` blocks that aren't needed often scatter through the codebase unnecessary.

```suggestion
pub struct FormatWrapper<T> {
    inner: T,
}

impl<T> FormatWrapper<T> {
    pub fn new(inner: T) -> Self {
        Self { inner }
    }
}
```

**Comment by thomaseizinger** [comment]
> Calls to the `render()` fn cannot be inlined.

The `Display` impl for `String` has an `#[inline]` attribute: https://doc.rust-lang.org/src/alloc/string.rs.html#2261
Perhaps we should also hint the compiler in the `Render` implementations to inline them?

**Comment by Pitasi** [comment]
> > Calls to the `render()` fn cannot be inlined.
> 
> The `Display` impl for `String` has an `#[inline]` attribute: [doc.rust-lang.org/src/alloc/string.rs.html#2261](https://doc.rust-lang.org/src/alloc/string.rs.html#2261) Perhaps we should also hint the compiler in the `Render` implementations to inline them?

I tried to sprinkle `#[inline]` here and there without any performance change 🤔 

Might be that FormatWrapper makes it impossible to statically dispatch the call to `.render()`, because it's generic and has no context of the concrete type being wrapped. I'm not a Rust expert though.

I didn't have much time this week to experiment further but one thing I can try is to get rid of FormatWrapper and call `::rscx::Renderable::render(thing)` directly from the macro, if I'm right this call can be inline since `thing` has a known concrete type. (It's similar to what we do with `EscapeAttribute` btw)

**Comment by Maaarcocr** [comment]
So I was able to work a bit on this and figured out where most of the time was going. The trait was badly implemented in some cases, but that did not give us that much of a boost when fixed. The real issue was being caused by an optimisation that the rust compiler was able to do when static strings where used in `format!` and was not being able to do anymore once we wrapped them. 

This is explained here: https://doc.rust-lang.org/std/fmt/struct.Arguments.html#method.as_str

So for this benchmark, we had something like: 

```rust
format!("<h1>{}{}</h1>", "my class is", { props.class })
```

rust is clever enough to figure out that `my class is` is a static string, so it just transforms the previous code to something like: 

```rust
format!("<h1>my class is{}</h1>", { props.class })
```

So how do we fix this? Well we also know when something is a static string (`RawText`) so we can just forcefully inline it in the `static_output` and sbam, we get all the performance back.

**Comment by Pitasi** [comment]
This is really good, thank you very much!

> rust is clever enough...

afaiu from your explanation Rust is NOT clever enough in this case, probably because there are two params and one of them is not a static string. That's where we lost all the performances.

From the doc you linked:

> The behavior for anything but the trivial case (without placeholders) is not guaranteed, and should not be relied upon for anything other than optimization.

Anyway, fixing this using `static_output` is definitely the right thing to do, thanks for investigating and gaining back our performance :)

**Comment by Pitasi** [comment]
I just released this in `v0.1.10` 🎉

**Comment by Maaarcocr** [comment]
for those following at home, small blog post about it, with a simpler reproduction of what was happening: https://procmarco.com/blog/2023-01-10-zero-cost/

**Pull request merged**